src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/CustomAttributeWrapper.cs

```diff
@@ -0,0 +1,185 @@
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection.Metadata;
+
+namespace System.Reflection.Emit
+{
+    internal readonly struct CustomAttributeWrapper
+    {
+        private readonly ConstructorInfo _constructorInfo;
+        private readonly byte[] _binaryAttribute;
+
+        public CustomAttributeWrapper(ConstructorInfo constructorInfo, ReadOnlySpan<byte> binaryAttribute)
+        {
+            _constructorInfo = constructorInfo;
+            _binaryAttribute = binaryAttribute.ToArray(); // TODO: Update to BlobHandle when public API public APi for MetadataBuilder.GetOrAddBlob(ReadOnlySpan<byte>) added
+        }
+
+        public ConstructorInfo Ctor => _constructorInfo;
+        public byte[] Data => _binaryAttribute;
+    }
+
+    internal struct CustomAttributeInfo
+    {
+        public ConstructorInfo _ctor;
+        public object?[] _ctorArgs;
+        public string[] _namedParamNames;
+        public object?[] _namedParamValues;
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2057:Unrecognized value passed to the parameter 'typeName' of method 'System.Type.GetType(String)'",
+            Justification = "The 'enumTypeName' only available at runtime")]
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2075:'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicFields', 'DynamicallyAccessedMemberTypes.NonPublicFields' in call to 'System.Type.GetField(String, BindingFlags)'",
+            Justification = "Could not propagate attribute into 'ctor.DeclaringType' only available at runtime")]
+        internal static CustomAttributeInfo DecodeCustomAttribute(ConstructorInfo ctor, ReadOnlySpan<byte> data)
+        {
+            int pos = 2;
+            CustomAttributeInfo info = default;
+
+            if (data.Length < 2)
+            {
+                throw new InvalidOperationException(SR.Format(SR.InvalidOperation_InvalidCustomAttributeLength, ctor.DeclaringType, data.Length));
+            }
+            if ((data[0] != 0x1) || (data[1] != 0x00))
+            {
+                throw new InvalidOperationException(SR.Format(SR.InvalidOperation_InvalidProlog, ctor.DeclaringType));
+            }
+
+            ParameterInfo[] pi = ctor.GetParameters();
+            info._ctor = ctor;
+            info._ctorArgs = new object?[pi.Length];
+            for (int i = 0; i < pi.Length; ++i)
+            {
+                info._ctorArgs[i] = DecodeCustomAttributeValue(pi[i].ParameterType, data, pos, out pos);
+            }
+            int numNamed = data[pos] + (data[pos + 1] * 256);
+            pos += 2;
+
+            info._namedParamNames = new string[numNamed];
+            info._namedParamValues = new object[numNamed];
+            for (int i = 0; i < numNamed; ++i)
+            {
+                int namedType = data[pos++];
+                int dataType = data[pos++];
+                string? enumTypeName = null;
+
+                if (dataType == 0x55)
```

Comment:
nit: use or add a enum or constants for data types

Reply:
Ah there is an Enum, but internal, maybe create a similar one https://github.com/dotnet/runtime/blob/c3a2bce1396441511e4a97f174e4d43c7648624f/src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs#L674

Reply:
End up using constants as not many values from the Enum is used pus Enum requires casting to int.